### ISSUE TITLE
Counter-examples to is-prime

### ISSUE THREAD

**Issue by trizen** [MoarVM/MoarVM#1170]
Just a comment on the current implementation of `is-prime($n)`, for large `$n`.

There seems to exist some numbers, published in [[1]](https://eprint.iacr.org/2018/749.pdf), that pass the primality test of LibTomMath, while being composite.

These are:

```perl
say is-prime(4755444900898962640570805079622407945513196432011911985648410022240088808985889200376095787527901239245756068158548164080646419498193938363223751878040975178651652955930683319090515589565114950046340315137578332784425652394857192999191244130079906060771897042318182261565231930157836276911534989784532752912947610193310744025493450293981924804993015182502555790249587996752315485361179895260300891395439822476185769523434215286135938029039660904526961670815522767096127450386003813203809365323894442795443565820952283705939380656438309180252554808691019535681908983623214664516034111470695703020642485875959651583213619466594932918476670651633816994599772548478768526764012575954414322984173288627);
say is-prime(6941543021392713730431668387068999032987505813628626223678951265009922979171494889619620919629218172938050022712650390231108995307237169269667644098333312684776469399924110119789527532161256933656112818109620945798457804143586629828296071826627003328849527952336389650813108369731049080747183836913592892951933560765345204468790419691959105305515726737868378312872715843312320064073674571001295106119373097331445689095722655847864191050899780290639922180983329597857624757187701943317347314318342630851888782724340309365489828447876503970224596096163190175586664838655752084811432643865453475409306234610422754846201955355349982116992680801631254169733179214516679698471486732095757632702985749004267668618407596194879010235811205069773345370151540294560707067912911972617117872568917574563797510424723636252899372760040898527772334935423922327351299656281932266256675132502702322760435460536571309302586539469310861406622510368133332971804029286468542260235396204271045794929827474636467550655521103657096119618759746982193439030809406498947869770588717242290031299573971411197982166717351757954266001066384132372994264058429334563291496657469558139808375030168417007373133166391119553220773091860698215980720042621281779381114805103592526045494482900527480395872621990534353634675251867);
say is-prime(106345207806296484231668043910066785399795652556314098984703724996175611913155370425059004666755351126370765757325069220154090475525113111331534155822504132427954121926028977897326016452032633264303162177531381326310998970713694521304856783427108211035837719173191888424497687716945903847840832601158220338315116311569353516661115745397618364437941836386218808101529448483687495168944238726821330736995943973596941803495501200742109710969089802656966120128959989924056907838239415947883597253487288911265130861076995785143322559232949628580082040985968381890859811544811963250166807225739883175758540592938038640525057077635260377283817570121188506720763322577723854603213904763949026451035886610117960695498777558294914095190286423434103595325322044355050505888390419892966130432853384420025991603809604285550077998959033403957562518661694503445321270371924756278334786001733438835926798599715925477942244937628383988529536076933049398733022131632892984494328799535905089231306547424707999668050846457897292105617153237859066140231207977706483096753560267776977308513770417672719215980272226316038096369457350381579229798547106736019854318870385754141371733858517502044218801634760265587498814280073328060552232702804872921969797735332629722017664747892121030708748712733180842486503682755515481434761547891191431690407378061063299887695616928512486921716155822944378708863947668462062195571745339003058779915940858554595239390172360173769656161537253021381577803596097455028941144286297184282046724577533468348343936096263977229393852616124982192888432604948830175933041671162863393593616389115022094081675889914163162852300506873558662214056728175728839204797446701664449699367321558065141345504373533192399722226922311408612622529245371708935701643178960931071329253928802738471610262504489056899252740549408250528391001108587552456472577625704595893048563596869581967781113019047808212452621728173827862405874506295664794063113108306419852891590636875366731732762664553725634243782507523770949879136684328870372364698244368287699980886402179469945231457642085062803806579094674257318340352731256782194724089323);
```

[[Try it online]](https://tio.run/##Zdc5rp5FEAXQnFUQQoDU8yBWQ@DAEkiWHbH6x7n9SBCyZf9Tdw13qPq@ffn@5/r4@PHH3z9//fHbt@9f//ryy9hzjjFuKeeeu9oaZe5yyizbuzb8N@asvd41eiu1Xn/PXOOMWkrLL8o5J6fnObeV0vcqd@6zZ9u31NZvG3PPVdap88xxqhinrLFGveOeevvtp6/eWt@znn3KKNerfdasa7Y75@1Od0mUW2ad89zpuzruLGWsPkqvctyi9t72GaP5XuRx5q633Xvl3caovSjoFsmUveu5u4zW66mntVXfGemUmotW20uxs8vxunO6WTK13bFXLX7Xa9ljlDbH7WOWphDVtKGA@25xrY/nnBrRJHvEXjtBtGH2Vet292yr9PS/9juHOzRVBKkv3/XRR6uznVW7PhxhSr9Lj8tw8q664KUnTb67eN@2ZPpZoDgVZs7cvqTfzwU1VCDeFXsaoFo7fZf5QCgLsLAot4ozZa6WowuqnU6oLkBrTW91rDXgU/qokNhFslsdDbJaoNZ554XSPH6s1uvABHjXw6NAyQbfLsl1kezevbXSV2evo7SNJ7XN3AM57GsX6XZv58Dm199/@g@R1x26Kr4uto0MvYxeF9ocgQ4KFJFlVaauLP1s@N3XBkBtC5HuFWHfuvFq4LKUlw7l39twZLe0aIb1@53oUO21aonulI2@KgkkSR2HkRFH2lKSD5YWC6FbUVHC4pPWAGz6DkL40FCxlheXaNQ7IwfFz7OkcdppF3URy9sN4XRj5KpQhYikPdEBNcF0t/9p@@CpM1HZDYsaoapaULLcqIFJZYyFn4JdPwJdLYoi/blB0YkRT45yzlb@PCOlYRe4y9ZGmTKEmnuLQm4AEL/XMRDnxhJojHThG@W7Po3bj9Crp/lIl565wsfYTwYuPXujX8LtPhQE1UMSq5dDYIdftN1GPODRfJx0aTwWBaINLuRaUcaKg1SOp5kcjEZIU@mHKYVgIb8b9GNP@DEd8qP1AfEwEyw6QwkxBWFqmAtfDc7d2FvXq3kjTMSxmCgQMeroYeSmIk4nKX65Y7sjRF/UffRxhnQjMFTUQtTatMmdLpXWjqfIjAUFOH8E0PLwiX9tfgKdtFuREFl9x0jFGNi5sJ6PsCYIcofxTLttsnP77XPEq2G6O0VwqkklgGzh2lQLi42fSb7HbOhrjoUlC0ppPsvQPq6Lzvi36ihONgl01hEx0FeNvRMUElC5zq5U2kN51iRXtB83VMdZ3u5nAGMVHK7mpvlMDktrLMZPAgQDpkaUoBuM5Kqhmb5R@9mkK5Bm99Q1N1bUuBa1O6uympK1jNMoFo0l3pkNz6T6O95YPIMB9HRVxweVzwSfsZOLAsaPBOqKR5ky@K9kl/ebSCjTmpnzvNX4SeDp2G6PAi3K35vBxEy5jErpNKNhGC9jnKQPLN/1DJIcYWmIqHkmYMYKOaz/@aJSnrwJjYFkbrdni9CjQXWSGvXtTFSGj8ZxrpDWJDe09HnHmRiHIdMzMFloiLtCCWJAUkSLj@hgGMIECSUKKlkv5OVsTTekm@mNEyXN/ZxFVSVhgSnLFd2wdDEpe9WfKOGm3M0u6xtpkuTWmd6Ix/gxg9vbSnacAYnOo4rJY8jTAtnHKchgXLNlx6d5gsmIJWTwQtbkxVBPdonIucaWU59CpR9BIQ7a4eSgrJFXDINvgZ@r2w4AxVeE4Mmo4L2pCywJ9R6fRD9epkpSzw3GJ27wTZQx51lm7P@N2cIuM0x8hx3Oqdj8KD6EZj/Zb3BICcdtdN85h3FwAwitRoFQCeE5JQPVdpzCXuRtJrSdo2XXOnJxn3WIpVphQBdDvCtgPq6w7qhGoKgkWxwKvHGofq9DCWCwtrfMcJTc56cyr9qDE4i@18vhGc6JdWQxCEdohBW27BNBjTtnRvLOLBXE7ERY40chp8FVYh9vNXo6oa0nr6w1MskfcPc3zDLq3JdtQM3gBYkRrpvZXfgFhfnCOfmnxeFFpkm8ej1NrUcXSGUHtF4pMGtevs7CGadhGbHo/qDg6OkpKCNhw5@BhHo24/QlfcxW2zN7OYoJ7SPHTSpC0fXkLM@blHeWCzllWlKumeTGGHoW2RtwfGOXntFOFhX8MAmy7oHdaI0NO6j99GVijOx3lJvZL1OgZbxkmMJNx2yRLfU9ezJaW@T2b6ue2HA3myW3fPsSJHHWeKOhvDGebVwANjK5bI@5endisCUYPb0oE@7lbQTEck6FQmxTyZbzrLNvTPKp88oiihRAkNQqbizG4NQLhUMFEcxDd7H7lhlMk1lEswSQHwJldAqTuF7t6ORmmyCnrDAxyJglv1cA4gV2zKi5RASm8eldN@pXIp72PAustxjEj20TNS7RM2pwMISKQkvKt1dkR4O1xFu6JebKlM3FPfrN8prGSqW/lcVlM8ZoWOnbWW9@eqGNUAIFGHk/1zsZgD1bDriySbe30ZNjFszxZmSLB8br3n7bswk9@nmCYHiul3ieMbIopTtCZw/lbejCV3QrvhjvaB5YiBpr3zKFuPGl3Z7oW7QnRXhZKncsJI@YnKhnC0oXs/rjRsrTzIAqvYyhFr5lC1FxDCxLOrG0bD/m6pFgLgQbAJWbZzATgc566kPYmLqaUStTxjpYshBllm45jtQTu6LEzL@s73lWmSIgU9zJOToAlmBvqgQLFDDo6WZnQutdfOtk0LQ8hqFHamfKzCAeffIQBq8ZimYn6fVzKS9Zfk/2sBp1ZlZwAn3Jk43Yeap6bCCQFqJJNo@MtJmnZfteD@Oy978JnQUJujiJ5eHtzTISV5ZTVrGMR0Pd3sn44sQtrg0AW6bveQlFZ6XGOd3L08gOWzQjGyIfsVJ8fPwD)

The prime factorization of these numbers is available on [factordb.com](http://factordb.com/index.php).

I suggest using the [Baillie–PSW primality test](https://en.wikipedia.org/wiki/Baillie%E2%80%93PSW_primality_test) instead, which is faster and has no known counter-examples.

[1] [Prime and Prejudice: Primality Testing Under Adversarial Conditions](https://eprint.iacr.org/2018/749.pdf).

**Comment by MasterDuke17** [comment]
Any chance you’re up for submitting a PR to libtommath so we get it for free when we next upgrade our libtommath version?

**Comment by trizen** [comment]
It looks like libtommath already has the B-PSW test implemented as `mp_prime_strong_lucas_selfridge(const mp_int *a, int *result)`.

See: https://github.com/libtom/libtommath/pull/113.

**Comment by MasterDuke17** [comment]
So if I'm reading those correctly, MoarVM should be fine once we upgrade to libtommath v1.1.0 (or greater)?

**Comment by jnthn** [comment]
@MasterDuke17 Well, perhaps we need to adapt our code to call the correct function also...

**Comment by MasterDuke17** [comment]
@jnthn in libtommath v1.1.0, `mp_prime_strong_lucas_selfridge` is called by `mp_prime_is_prime`, while it doesn't even exist in the version MoarVM is using (v1.0). So I think all that's needed is to update our libtommath.

**Comment by jnthn** [comment]
@MasterDuke17 Ah, I'd missed that. So yes, sounds good. Thanks!

**Comment by trizen** [comment]
Seems to be fixed in MoarVM version 2020.05.
